### ISSUE TITLE
Add type checking with Pyright to CI gates

### ISSUE THREAD

**Issue by vytas7** [falconry/falcon#2373]
At the time of this writing, we are only checking with Mypy.
Also run the same checks with Pyright.

**Comment by dmaaskant** [comment]
Hello,

Out of curiosity, I was wondering what you are referring to by CI gates. I can't seem to find any information on this online. 

Would you be able to help me understand this concept?

**Comment by vytas7** [comment]
Hey @dmaaskant!
Sure, by the acronym _CI_ I was referring to [**C**ontinuous **I**ntegration](https://en.wikipedia.org/wiki/Continuous_integration). By _CI gates_ I was referring to CI jobs, processes, etc, that must pass before a pull request can be merged into our main branch; see also here: [What is a gating continuous integration (CI) system?](https://devops.stackexchange.com/questions/10600/what-is-a-gating-continuous-integration-ci-system)

TL;DR We're looking into potentially integrating Pyright checks into our [GitHub Actions](https://github.com/falconry/falcon/actions) setup.
